sunscreen/src/types/zkp/bfv_plaintext.rs

```diff
@@ -0,0 +1,271 @@
+use sunscreen_compiler_macros::TypeName;
+use sunscreen_runtime::ShareWithZkp;
+use sunscreen_zkp_backend::{BigInt, FieldSpec};
+
+use crate::{
+    invoke_gadget,
+    types::{bfv::Signed, zkp::ProgramNode},
+    zkp::{with_zkp_ctx, ZkpContextOps},
+};
+
+use super::{gadgets::SignedModulus, DynamicNumFieldElements, Field, ToNativeFields};
+
+use crate as sunscreen;
+
+/// A BFV plaintext polynomial that has been shared to a ZKP program.
+///
+/// Here `N` is the degree of the _shared_ polynomial, which may be less than the full lattice
+/// dimension. See [`Share::DEGREE_BOUND`](sunscreen_runtime::Share).
+#[derive(Debug, Clone, TypeName)]
+struct BfvPlaintext<F: FieldSpec, const N: usize> {
+    data: Vec<Field<F>>,
+}
+
+impl<F: FieldSpec, const N: usize> DynamicNumFieldElements for BfvPlaintext<F, N> {
+    fn num_native_field_elements(plaintext_modulus: u64) -> usize {
+        let log_p = plaintext_modulus.ilog2() + 1;
+        log_p as usize * N
+    }
+}
+
+impl<F: FieldSpec, const N: usize> ToNativeFields for BfvPlaintext<F, N> {
+    fn to_native_fields(&self) -> Vec<BigInt> {
+        self.data.iter().map(|x| x.val).collect()
+    }
+}
+
+#[derive(Debug, Clone, TypeName)]
+/// A [BFV signed integer](crate::types::bfv::Signed) that has been shared to a ZKP program.
+///
+/// Use the [`AsFieldElement::into_field_elem`] method to decode the the value into a field
+/// element within a ZKP program.
+pub struct BfvSigned<F: FieldSpec>(BfvPlaintext<F, { <Signed as ShareWithZkp>::DEGREE_BOUND }>);
+
+impl<F: FieldSpec> DynamicNumFieldElements for BfvSigned<F> {
+    fn num_native_field_elements(plaintext_modulus: u64) -> usize {
+        <BfvPlaintext<F, { <Signed as ShareWithZkp>::DEGREE_BOUND }> as DynamicNumFieldElements>::
+            num_native_field_elements(plaintext_modulus)
+    }
+}
+
+impl<F: FieldSpec> ToNativeFields for BfvSigned<F> {
+    fn to_native_fields(&self) -> Vec<BigInt> {
+        self.0.to_native_fields()
+    }
+}
+
+/// Decode the underlying plaintext polynomial into the field.
+pub trait AsFieldElement<F: FieldSpec> {
+    /**
+     * Return a structure scaled by `x`.
+     */
```

Comment:
Is this comment correct? Or is it something that converts the field elements into a signed field element?

Reply:
I have no idea where this came from. Copilot? :smile: